CompilerSource/parsing/declarator.cpp

```diff
@@ -0,0 +1,284 @@
+/** Copyright (C) 2022 Dhruv Chawla
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+#include "declarator.h"
+
+#include "ast.h"
+#include "full_type.h"
+
+namespace enigma::parsing {
+FunctionParameterNode::Parameter::Parameter(bool is_variadic, void *default_value, std::unique_ptr<FullType> type):
+  is_variadic{is_variadic}, default_value{default_value}, type{std::move(type)} {}
+FunctionParameterNode::Parameter::~Parameter() {
+  delete reinterpret_cast<AST::Initializer *>(default_value);
+}
+
+FunctionParameterNode::FunctionParameterNode(std::vector<Parameter> params, bool outside_nested):
+  kind{Kind::DECLARATOR}, parameters{std::move(params)}, outside_nested{outside_nested} {}
+FunctionParameterNode::FunctionParameterNode(Parameters params, bool outside_nested):
```

Comment:
`outside_nested` is basically my jank ass solution to make converting stuff like `*(*x)[10]` to an expression easier because the `[10]` is stored as a part of the `(*x)` in the parser. This is because `*(*x)[10]` needs to be interpreted as `(*(*x))[10]` however the `to_expression` thing considers trailing modifiers (array indices, function calls) before pointer/reference modifiers, so it would instead lead to something like `*((*x)[10])`, which is semantically wrong. The `outside_nested` flag makes `to_expression` consider trailing modifiers after the leading modifiers